src/crypto/sharing.cpp

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "sharing.h"
+
+#include "ccf/crypto/entropy.h"
+
+#include <stdexcept>
+
+namespace crypto
+{
+  /* PRIME FIELD
+
+     For simplicity, we use a finite field F[prime] where all operations
+     are defined in plain uint64_t arithmetic, and we reduce after every
+     operation. This is not meant to be efficient. Compared to e.g. GF(2^n), the
+     main drawback is that we need to hash the "raw secret" to obtain a
+     uniformly-distributed secret.
+  */
+
+  typedef uint64_t element;
+
+  const uint64_t prime = (1ul << 31) - 1ul; // a notorious Mersenne prime
+
+  element reduce(uint64_t x)
+  {
+    return (x % prime);
+  }
+
+  element mul(element x, element y)
+  {
+    return ((x * y) % prime);
+  }
+
+  element add(element x, element y)
+  {
+    return ((x + y) % prime);
+  }
+
+  element sub(element x, element y)
+  {
+    return ((prime + x - y)) % prime;
+  }
+
+  // naive algorithm, used only to compute coefficients, not for use on secrets!
+  element exp(element x, size_t n)
+  {
+    element y = 1;
+    while (n > 0)
+    {
+      if (n & 1)
+        y = mul(y, x);
+      x = mul(x, x);
+      n >>= 1;
+    }
+    return y;
+  }
+
+  element inv(element x)
+  {
+    if (x == 0)
+    {
+      throw std::invalid_argument("division by zero");
+    }
+    return exp(x, prime - 2);
+  }
+
+  // This function is specific to prime=2^31-1.
+  // We assume the lower 31 bits are uniformly distributed,
+  // and retry if they are all set to get uniformity in F[prime].
+
+  element sample(const crypto::EntropyPtr& entropy)
+  {
+    uint64_t res = prime;
+    while (res == prime)
+    {
+      res = entropy->random64() & prime;
+    }
+    return res;
+  }
+
+  /* POLYNOMIAL SHARING AND INTERPOLATION */
+
+  void sample_polynomial(
+    element p[], size_t degree, const crypto::EntropyPtr& entropy)
```

Comment:
```suggestion
    element* p, size_t degree, const crypto::EntropyPtr& entropy)
```

Reply:
That seems worse if anything, but I can make it a span. It's completely internal, so I'm not too worried.